### PR TITLE
Github Actions documentation/opt. and AVX2-detection changes

### PR DESCRIPTION
**Pull request by fabratu**
As discussed in #674 (and in preperation for the distribution of a wheel-package) this PR reduces the total number of tests slightly. This also changes:

- Remaining tests are renamed according to a certain scheme, which is now documented at the beginning of the `ci.yml` file. This also gives a rationale, why each test is done.
- ~~Shortening the name of each test, so that the full name is visible in the Actions-overview for a run. Hopefully the emojis are not to fancy~~
- Remove duplicating lines of code in `ci.myl` by using a "[fake ternary operator](https://github.community/t/do-expressions-support-ternary-operators-to-change-their-returned-value/18114)"
- Removing custom tests in main `CMakeLists.txt` for AVX2. `-march=native` already should enable all hardware-specific features like AVX2.
- Added custom-test for Windows/MSVC in order to detect AVX2-support. `-march=native` or similar is not available in MSVC.

Any input is appreciated :) Choosing the "right" test-cases is always depending on ones point of view.

### CODE REVIEW

**Comment by avdgrinten** [comment]
The Emoji are too confusing to people who don't know NetworKit's CI infrastructure.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,7 +1,104 @@
...
+#              C++17 = Check for C++17 conformance
+#              C++20 = Check for C++20 conformance
+# 
+#
+# clang-tidy
+# ----------
+# Description: Uses clang-tidy for several static analysis checks. These involve bug prone programming
+#              errors, possible modernizations and performance optiomizations.
+# Script:      scripts/clang_tidy.sh
+# Config:      /.clang-tidy
```
Comment (angriman) [inline .github/workflows/ci.yml:44]: Remove the `/`?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,7 +1,104 @@
...
+# Build-type:  Debug
+#
+# 
+# coverage
+# --------
+# Description: Enables code-coverage, leak-sanitizers and tests in order to create cpp-coveralls data.
```
Comment (angriman) [inline .github/workflows/ci.yml:50]: Maybe specify address and leak sanitizers (address sanitizers are enabled by default if one sets `-DNETWORKIT_WITH_SANITIZERS=leak`).

**Comment by angriman** [comment]
Some build names are called something like `Linux (gcc-10)` and others `Linux (llvm-11)`. However, llvm is not just a compiler, the name should be changed to something like `Linux (llvm clang-11)` or just `Linux (clang-11)`.

**Comment by fabratu** [comment]
Adressed all comments.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,7 +1,104 @@
...
+# Description: Builds the C++ code into one shared library. Has also sanity checks enabled.
+# Script:      scripts/core.sh
+# Build-type:  Debug
+#
+# 
+# core (variant)
+# ---------------------
+# Description: These builds are variants of core with different options. 
+# Script:      scripts/core.sh
+# Variants:    release = Release-build
```
Comment (avdgrinten) [inline .github/workflows/ci.yml:33]: Why does `core` need a `release` variant?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -1,7 +1,104 @@
...
+#              non-monolithic = Each C++ module is a library
+#              C++17 = Check for C++17 conformance
+#              C++20 = Check for C++20 conformance
```
Comment (avdgrinten) [inline .github/workflows/ci.yml:36]: I think the C++17 build can probably be dropped. I know it has existed for quite some time, but if C++20 works, there is little reason that C++17 wouldn't.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -1,7 +1,104 @@
...
+#              native = Build for native architecture
```
Comment (avdgrinten) [inline .github/workflows/ci.yml:37]: Why do we need a native core build if we also have a native full build?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -1,7 +1,104 @@
...
+# full (variant)
+# ---------------------
+# Description: These builds are variants of full with different options.
+# Script:      scripts/full.sh
+# Variants:    native = Build for native architecture
+# 
+#         
+# core
+# ----
+# Description: Builds the C++ code into one shared library. Has also sanity checks enabled.
```
Comment (avdgrinten) [inline .github/workflows/ci.yml:24]: This description misses the rationale for `core`: the reason why we do these core builds is that they are faster than `full` builds, right?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -1,7 +1,104 @@
...
+# ----
+# Description: Basic build with all components (TLX, C++, Cython). Has to be tested on every fully 
+#              supported platform. Close to user-experience, except that warnings are also treated 
+#              as errors during build and TLX is provided as externally build library (normal 
+#              installation builds both NetworKit and TLX in one step).
+# Script:      scripts/full.sh
+# Build-type:  Release
+#
+# 
+# full (variant)
```
Comment (avdgrinten) [inline .github/workflows/ci.yml:15]: Merge the variants into the description of `full`.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -1,7 +1,104 @@
...
+# 
+#         
+# core
+# ----
+# Description: Builds the C++ code into one shared library. Has also sanity checks enabled.
+# Script:      scripts/core.sh
+# Build-type:  Debug
+#
+# 
+# core (variant)
```
Comment (avdgrinten) [inline .github/workflows/ci.yml:29]: Merge the variants into the description of `core`.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -1,7 +1,104 @@
...
+#              when a pull request is merged.
+# Script:      scripts/documentation.sh 
+# Build-type:  Debug
+#
+#
+# Jobs:
+# =====
+# 0. Rationale:
+# - Main platform is Linux, which is therefore used for most of the tests. Basic user experience (full) 
+#   should be tested on all platforms  with the default compiler.
```
Comment (avdgrinten) [inline .github/workflows/ci.yml:68]: What does "on all platforms" mean here if we previously only refer to Linux?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (avdgrinten) [reply .github/workflows/ci.yml:33]: If there is a reason, state it in the description. (I guess it is only used for Windows?)

File: .github/workflows/ci.yml (hunk 6 above)
Comment (fabratu) [reply .github/workflows/ci.yml:24]: Yes, main argument is heaving more speed for Debug-builds. `full` is by default a Release-build, while `core` is a Debug-build. Added this information to the description.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (fabratu) [reply .github/workflows/ci.yml:68]: Should refer to all supported platforms. Restructured this point to make it clearer.

**Review by avdgrinten — APPROVED** [review]
LGTM now.

**Pull request merged**